In [1]:
#Import libraries 
import pandas as pd 
import numpy as np 
import os 

## Working through 4.5 lesson

In [2]:
#Path to Instacart Basket Analysis Project 
path = r'/Users/libertyonia/Documents/Instacart Basket Analysis' 

In [3]:
#Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [4]:
#Import orders_wrangled.csv 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [5]:
#Create a dataframe 
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column 
df_test['mix'] = ['a','b',1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
#Check mixed datatypes
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
#Change numeric datatypes in the mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
#Check if df_prods has any missing values 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
#Checks for observations where there are missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
df_prods.shape

(49693, 5)

In [12]:
#Create a subset containing no missing values in the product_names column 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

In [13]:
#Creates a new subset containing only rows that are duplicates 
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups 

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
#Drop duplicate rows in df_prods_clean dataframe 
df_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_no_dups.shape

(49672, 5)

In [17]:
#Export clean dataframe 
df_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_no_duplicates_4.5.csv'))

# 4.5 Tasks 

In [18]:
#Check statistical descriptions 
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### Answer #2. I do not see any data inconsistencies when observing the descriptive statistic values of the df_ords dataset. The 'order day of the week' column has a min of 0 and a max of 6 which cover 7 days of the week. The 'order hour of the day' column has a min of 0 and max of 23 which covers 24 hours of the day. 

In [19]:
#3. Check for mixed-type data in your df_ords dataframe 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Answer #3-4. There are no columns that have mixed-type data

In [20]:
#5. Looks for missing values in the df_ords data frame 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Answer #5. The 'days_since_prior_order' column has 206,209 missing values. Missing values may be caused by the fact that first time customers will have NaN prior days since order. 

In [21]:
#6. Address the missing values using an appropriate method 
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]
df_ords_clean.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_day_of_week         0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

### Answer #6. The missing values in the 'days_since_prior_order' column covers 6% of all data records in this column. Instead of overwriting the df_orders dataframe, I decided to reassign the clean dataframe to df_ords_clean. 

In [22]:
#7. Check for duplicate values in the df_ords_clean dataframe 
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### Answer #7. There are no duplicate rows in the df_ords_clean dataset  

### Answer #8. There are no duplicate rows in the df_ords_clean dataset, so I will not do anything to remove any duplicates. 

In [23]:
#Export cleaned df_ords data set 
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'ords_cleaned_4.5.csv'))